Bluetooth: Host: Deprecate `BT_LE_ADV_OPT_USE_NAME` and `BT_LE_ADV_OPT_FORCE_NAME_IN_AD`

theob-pro opened the pull request:
Deprecate `BT_LE_ADV_OPT_USE_NAME` and `BT_LE_ADV_OPT_FORCE_NAME_IN_AD`. Update the code-base to not use those options anymore.

Reasons:

- It is out of place in the advertiser options, as it acts on advertising data and not configuration
- The behavior is not clear. All of these are trick questions:
  - where does the data for the name go?
  - do I have to allocate memory for this?
  - does it work the same in legacy vs extended advertising?
- The trade-off between UX improvement and extra complexity in the host is debatable. Hundreds of lines in the stack to avoid making the user type out less than ten.

See commits message for details.

theob-pro committed:
Bluetooth: Host: Deprecate auto name in AD

Deprecate the advertiser options `BT_LE_ADV_OPT_USE_NAME` and
`BT_LE_ADV_OPT_FORCE_NAME_IN_AD`.

Reasons:

- It is out of place in the advertiser options, as it acts on
  advertising data and not configuration
- The behavior is not clear. All of these are trick questions:
  - where does the data for the name go?
  - do I have to allocate memory for this?
  - does it work the same in legacy vs extended advertising?
- The trade-off between UX improvement and extra complexity in the host
  is debatable. Hundreds of lines in the stack to avoid making the user
  type out less than ten.

Also deprecate the following macros that uses those options:

- `BT_LE_ADV_CONN_NAME`
- `BT_LE_ADV_CONN_NAME_AD`
- `BT_LE_ADV_NCONN_NAME`
- `BT_LE_EXT_ADV_CONN_NAME`
- `BT_LE_EXT_ADV_SCAN_NAME`
- `BT_LE_EXT_ADV_NCONN_NAME`
- `BT_LE_EXT_ADV_CODED_NCONN_NAME`

Update the migration guide and the release note for Zephyr 3.7 to
describe those changes.

Signed-off-by: Théo Battrel <theo.battrel@nordicsemi.no>

theob-pro committed:
Bluetooth: Host: Add new advertiser params macro

With the deprecation of the advertiser options to automatically add the
device name in the scan response and advertising data, some macro used
to set the advertiser parameter also became deprecated. That's the case
for `BT_LE_EXT_ADV_CONN_NAME` and `BT_LE_EXT_ADV_SCAN_NAME`. The issue
is that those two macro didn't had their equivalent without including
the name. So this commit add `BT_LE_EXT_ADV_CONN` and
`BT_LE_EXT_ADV_SCAN` to be used instead.

Signed-off-by: Théo Battrel <theo.battrel@nordicsemi.no>

theob-pro committed:
Bluetooth: Tests: Remove usage of auto name in AD

The advertiser options to automatically add the name in the scan
response and the advertising data will be deprecated.

Update the tests that were relying on those options. A lot of test were
actually not using the name, for those the name have simply been
removed. For the others that were using the name, add it in the
advertising data or the scan response data.

Signed-off-by: Théo Battrel <theo.battrel@nordicsemi.no>

theob-pro committed:
Bluetooth: Samples: Remove uses of auto name in AD

The advertiser options to automatically add the name in the scan
response and the advertising data will be deprecated.

Update the samples that were using those options by explicitly adding
the device name in the advertising data or the scan response data.

Signed-off-by: Théo Battrel <theo.battrel@nordicsemi.no>

theob-pro committed:
Bluetooth: Shell: Remove usage of auto name in AD

The advertiser options to automatically add the name in the scan
response and the advertising data will be deprecated.

Update the Bluetooth shell and its documentation to not use those
options anymore.

The option of the legacy advertiser run with `bt advertise` and the
behavior hasn't changed. The code has been updated to no longer use the
advertiser option but the original behavior has been reproduced.

The option of the extended advertiser have been updated to be more
explicit. The options `name` and `name-ad` have been removed from the
`bt adv-create` and `bt adv-param`. The option `name` and `dev-name`
have been added to the `bt adv-data` command.

- `name <str>` will advertise the given name.
- `dev-name` will advertise the device name.

Signed-off-by: Théo Battrel <theo.battrel@nordicsemi.no>

theob-pro committed:
Bluetooth: Mesh: Remove usage of auto name in AD

The advertiser options to automatically add the name in the scan
response and the advertising data will be deprecated.

The Mesh files `proxy_srv.c` and `pb_gatt_srv.c` were using
BT_LE_ADV_OPT_USE_NAME. Add the device name in the advertising data
directly.

Signed-off-by: Théo Battrel <theo.battrel@nordicsemi.no>

alxelax commented on subsys/bluetooth/mesh/proxy_srv.c:
```diff
@@ -487,6 +481,9 @@ static const struct bt_data net_id_ad[] = {
 	BT_DATA_BYTES(BT_DATA_UUID16_ALL,
 		      BT_UUID_16_ENCODE(BT_UUID_MESH_PROXY_VAL)),
 	BT_DATA(BT_DATA_SVC_DATA16, proxy_svc_data, NET_ID_LEN),
+#if defined(CONFIG_BT_MESH_PROXY_USE_DEVICE_NAME)
```
I guess, the same should be added to `enc_id_ad`.

theob-pro replied:
You're right, thanks for the catch. I added it.

theob-pro commented on subsys/bluetooth/host/data.c:
```diff
@@ -78,3 +78,15 @@ size_t bt_data_serialize(const struct bt_data *input, uint8_t *output)
+
+	ad.data = (const uint8_t *) device_name;
+	ad.type = BT_DATA_NAME_COMPLETE;
+	ad.data_len = strlen(device_name);
```
The problem is more with `bt_get_name`. I am not even sure that the function I am adding is that useful. I will maybe just remove it.
Anyway I think fixing the race between `bt_get_name` and `bt_set_name` should be done in another PR.

jori-nordic replied:
If you're ok w/ it (@alwa-nordic) I second theo in de-scoping that API from this PR.
We want to minimize discussion here as the rebase is quite tedious.

@theob-pro If the only thing using that API is the tests, just implement a version of `bt_get_name` in the testlib with a sched-lock. We can discuss the proper solution in the next PR.

theob-pro commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -1833,18 +1849,30 @@ static int cmd_advertise(const struct shell *sh, size_t argc, char *argv[])
 		}
 	}
 
+	if (name_ad && !no_name) {
```
You're right, it's better. I wanted to stick to the command option name but that double negation is really bad 😅

theob-pro commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -1833,18 +1849,30 @@ static int cmd_advertise(const struct shell *sh, size_t argc, char *argv[])
+	}
+
+	if (name_sd && !no_name) {
+		set_ad_device_name_complete(&sd[0]);
```
I can't find anything saying you're not allowed to have both but anyway if you look in the loop, when `name-ad` is selected `name_sd` is set to false, so you cannot have both with that implementation

theob-pro commented:
Rebased to fix failing tests and apply change requested by comments.
- `set_ad_device_name_complete` function has been removed following @alwa-nordic comment;
- @alxelax I updated the Mesh file to put the name in the scan response data instead of the advertising data, it was the actual behavior of `BT_LE_ADV_OPT_USE_NAME`

theob-pro commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -1762,14 +1762,33 @@ static ssize_t ad_init(struct bt_data *data_array, const size_t data_array_size,
+	ad->data = name;
+}
+
+void set_ad_device_name_complete(struct bt_data *ad)
```
Removed the function, no more naming problem :smile:

Thalley commented on include/zephyr/bluetooth/bluetooth.h:
```diff
@@ -549,25 +549,29 @@ enum {
-	 *  BT_DATA(BT_DATA_NAME_COMPLETE, name, sizeof(name) - 1)
-	 *  @endcode
+	/**
+	 * \deprecated This option will be removed in the near future, see
```
Generally we use `@` instead of `\` in doxygen although the should result in the same 

```suggestion
	 * @deprecated This option will be removed in the near future, see
```

Thalley commented on include/zephyr/bluetooth/bluetooth.h:
```diff
@@ -549,25 +549,29 @@ enum {
-	 *  @code
-	 *  BT_DATA(BT_DATA_NAME_COMPLETE, name, sizeof(name) - 1)
-	 *  @endcode
+	/**
+	 * \deprecated This option will be removed in the near future, see
+	 * https://github.com/zephyrproject-rtos/zephyr/issues/71686
+	 *
+	 * @brief Advertise using GAP device name.
```
```suggestion
	/** @brief Advertise using GAP device name.
	 * 
	 *  @deprecated This option will be removed in the near future, see
	 *  https://github.com/zephyrproject-rtos/zephyr/issues/71686
	 *
	 * 
```

I would revert back to this and add deprecated below the @brief to avoid this many line changes.

theob-pro replied:
I think it's nice to have it at the top. In the documentation you instantly see that the macro is deprecated.

Thalley commented on include/zephyr/bluetooth/bluetooth.h:
```diff
@@ -691,10 +695,13 @@ enum {
 	 *
-	 * Will place the GAP device name into the advertising data rather
-	 * than the scan response data.
+	 * Will place the GAP device name into the advertising data rather than
+	 * the scan response data.
```
This change seems unnecessary

theob-pro replied:
My formatter did it, I don't think that's a big issue. Especially knowing that those options will be deleted in the future :smile:

Thalley commented on samples/bluetooth/direct_adv/src/main.c:
```diff
@@ -76,6 +76,7 @@ BT_GATT_SERVICE_DEFINE(primary_service,
 static const struct bt_data ad[] = {
 	BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
 	BT_DATA_BYTES(BT_DATA_UUID128_ALL, BT_UUID_CUSTOM_SERVICE_VAL),
+	BT_DATA(BT_DATA_NAME_COMPLETE, CONFIG_BT_DEVICE_NAME, sizeof(CONFIG_BT_DEVICE_NAME) - 1),
```
Would it make sense to do
```suggestion
	BT_DATA(BT_DATA_NAME_COMPLETE, CONFIG_BT_DEVICE_NAME, strlen(CONFIG_BT_DEVICE_NAME)),
```

Thalley replied:
Not sure if too specific, but we could considering introducing a `BT_DATA_NAME` macro that does this, so it would just be 

```c

static const struct bt_data ad[] = {
	BT_DATA_NAME,
};
```

theob-pro replied:
I prefer to keep the explicit definition, we are trying to remove magic, let's not hide thing again

Thalley commented on tests/bluetooth/common/testlib/src/adv.c:
```diff
@@ -35,7 +35,7 @@ static void connected_cb(struct bt_le_ext_adv *adv, struct bt_le_ext_adv_connect
 }
 
-int bt_testlib_adv_conn(struct bt_conn **conn, int id, uint32_t adv_options)
+int bt_testlib_adv_conn(struct bt_conn **conn, int id)
```
Doesn't it make sense to keep the options here? Since this is a library, tests may still want to set other options.

theob-pro replied:
@alwa-nordic what do you think, IMO it's not useful to have options here as it's a very specific function used to create a connection, not just start an advertiser

omkar3141 reviewed (dismissed):
Mesh changes look ok.

Thalley commented on tests/bluetooth/common/testlib/src/adv.c:
```diff
@@ -60,6 +60,16 @@ int bt_testlib_adv_conn(struct bt_conn **conn, int id, uint32_t adv_options)
+
+		ad.type = BT_DATA_NAME_COMPLETE;
+		ad.data_len = strlen(dev_name);
+		ad.data = (const uint8_t *) dev_name;
```
```suggestion
		ad.data = (const uint8_t *)dev_name;
```

Thalley commented on tests/bluetooth/tester/src/btp_bap_broadcast.c:
```diff
@@ -347,8 +351,11 @@ uint8_t btp_bap_broadcast_source_setup(const void *cmd, uint16_t cmd_len,
-					     0, &gap_settings);
+	sd.type = BT_DATA_NAME_COMPLETE;
+	sd.data_len = strlen(dev_name);
+	sd.data = (const uint8_t *) dev_name;
```
```suggestion
	sd.data = (const uint8_t *)dev_name;
```

We generally do not have spaces between the casting type and the value

Thalley commented on tests/bluetooth/common/testlib/src/adv.c:
```diff
@@ -60,6 +60,16 @@ int bt_testlib_adv_conn(struct bt_conn **conn, int id, uint32_t adv_options)
 	g_ctx = &ctx;
 
 	api_err = bt_le_ext_adv_create(&param, &cb, &adv);
+	if (!api_err) {
+		struct bt_data ad;
+
+		ad.type = BT_DATA_NAME_COMPLETE;
+		ad.data_len = strlen(dev_name);
+		ad.data = (const uint8_t *) dev_name;
```
Should we add another argument so that the caller can control whether or not to set the name in ad/sd?

theob-pro replied:
As said previously this function is very specific, I don't think the user should care about that ^^

Thalley reviewed (changes requested):
A few comments. 

W.r.t. changing a stable API like we do here, do we need to go through the steps in https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#introducing-breaking-api-changes?

theob-pro commented:
> A few comments.
> 
> W.r.t. changing a stable API like we do here, do we need to go through the steps in https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#introducing-breaking-api-changes?

IMO it's not a breaking API change, it's a deprecation: https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#deprecated

Thalley commented:
> > A few comments.
> > W.r.t. changing a stable API like we do here, do we need to go through the steps in https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#introducing-breaking-api-changes?
> 
> IMO it's not a breaking API change, it's a deprecation: https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#deprecated

You are completely right :) Thanks

Thalley commented on tests/bluetooth/common/testlib/src/adv.c:
```diff
@@ -60,6 +59,16 @@ int bt_testlib_adv_conn(struct bt_conn **conn, int id, uint32_t adv_options)
 	g_ctx = &ctx;
 
 	api_err = bt_le_ext_adv_create(&param, &cb, &adv);
+	if (!api_err) {
+		struct bt_data ad;
+
+		ad.type = BT_DATA_NAME_COMPLETE;
+		ad.data_len = strlen(name);
+		ad.data = (const uint8_t *)name;
+
+		api_err = bt_le_ext_adv_set_data(adv, &ad, 1, NULL, 0);
+	}
```
What if `name` is `NULL`? Should this be
```suggestion
	if (!api_err && name != NULL) {
		struct bt_data ad;

		ad.type = BT_DATA_NAME_COMPLETE;
		ad.data_len = strlen(name);
		ad.data = (const uint8_t *)name;

		api_err = bt_le_ext_adv_set_data(adv, &ad, 1, NULL, 0);
	}
```

theob-pro replied:
You're right, sorry I have been too quick 😅
It's updated

jhedberg reviewed (changes requested):
@theob-pro maybe I missed it, but I can't find the place where you provide a justification for the change, i.e. answering the question *why* these are deprecated. I'm not saying I'm against the actual change, but a clear justification is one of the most important things that needs to be documented (e.g in commit messages, PR description, code comments, etc).

theob-pro commented:
> @theob-pro maybe I missed it, but I can't find the place where you provide a justification for the change, i.e. answering the question _why_ these are deprecated. I'm not saying I'm against the actual change, but a clear justification is one of the most important things that needs to be documented (e.g in commit messages, PR description, code comments, etc).

You're right, I'll update the message of the commit doing the deprecation, the issue tracking the deprecation and the PR message.

alxelax reviewed (approved):
Mesh related changes are Ok

jori-nordic reviewed (approved):
🚂✅✅✅✅

sjanc commented on tests/bluetooth/tester/src/btp_bap_broadcast.c:
```diff
@@ -347,8 +351,11 @@ uint8_t btp_bap_broadcast_source_setup(const void *cmd, uint16_t cmd_len,
+	sd.type = BT_DATA_NAME_COMPLETE;
+	sd.data_len = strlen(dev_name);
+	sd.data = (const uint8_t *)dev_name;
+	err = tester_gap_create_adv_instance(param, BTP_GAP_ADDR_TYPE_IDENTITY, &base_ad, 1,
```
Hmm providing both AD and SD is valid only for instances using legacy PDUs, isin't ?

theob-pro replied:
I think you're right, I made a mistake here and on `btp_cap.c`. I'll create a PR to fix that in the afternoon.
Thanks for catching that! 🙂

sjanc replied:
I think some other samples that use ext adv has similar issue,  eg periodic_adv sample where SD data is provided to non-scannable ext adv instance

I'd suggest to re-review those samples where non-legacy instances are used

theob-pro replied:
I'll do it, that's a bit sad that not all samples are tested in CI 😅

theob-pro replied:
Created the PR: https://github.com/zephyrproject-rtos/zephyr/pull/71848
I looked at the other samples and did not found anything
